pkg/mimirtool/client/backfill.go

```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/fs"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+)
+
+func (c *MimirClient) Backfill(ctx context.Context, source string, logger log.Logger) error {
+	// Scan blocks in source directory
+	es, err := os.ReadDir(source)
+	if err != nil {
+		return errors.Wrapf(err, "failed to read directory %q", source)
+	}
+	var failed []string
+	var succeeded []string
+	var alreadyExists []string
+	for _, e := range es {
+		pth := filepath.Join(source, e.Name())
+		if err := c.backfillBlock(ctx, pth, logger); err != nil {
+			if errors.Is(err, errConflict) {
+				level.Warn(logger).Log("msg", "failed uploading block since it already exists on server",
+					"path", pth)
+				alreadyExists = append(alreadyExists, pth)
+			} else {
+				level.Warn(logger).Log("msg", "failed uploading block", "path", pth, "err", err)
+				failed = append(failed, pth)
+			}
+			continue
+		}
+
+		level.Info(logger).Log("msg", "successfully uploaded block", "path", pth)
+		succeeded = append(succeeded, pth)
+	}
+
+	level.Info(logger).Log("msg", "finished uploading block(s)", "succeeded", len(succeeded),
+		"already_exists", len(alreadyExists), "failed", len(failed))
+
+	if len(failed) > 0 {
+		return fmt.Errorf("failed to upload %d block(s)", len(failed))
+	}
+
+	return nil
+}
+
+func closeResp(resp *http.Response) {
+	// Drain and close the body to let the transport reuse the connection
+	io.Copy(io.Discard, resp.Body)
+	resp.Body.Close()
+}
+
+func (c *MimirClient) backfillBlock(ctx context.Context, dpath string, logger log.Logger) error {
```

Comment:
Not sure what `dpath` stands for. Can we rename it `blockDir` (also in `getBlockMeta`)?

Reply:
Done.